combinetf2/fitter.py

```diff
@@ -980,167 +972,44 @@ def fun():
             exp = self.expected_variations(fun, correlations=correlated_variations)
         else:
             exp = tf.function(fun)()
 
         if compute_chi2:
+            data, data_var, data_cov = model.get_data(self.nobs, self.data_cov_inv)
 
-            def fun_residual():
-                return (
-                    self._compute_yields(
-                        inclusive=inclusive,
-                        profile=profile,
-                        full=False,
-                    )
-                    - self.nobs
-                )
-
-            if profile:
-                res, resvar, rescov, _1, _2 = self._expvar_profiled(
-                    fun_residual, compute_cov=True
-                )
-            else:
-                res, resvar, rescov, _1, _2 = self._expvar(
-                    fun_residual, compute_cov=True
+            # need to calculate prediction excluding masked channels
+            def flat_fun():
+                return self._compute_yields(
+                    inclusive=inclusive, profile=profile, full=False
                 )
 
-            chi2val = self.chi2(res, rescov).numpy()
-            ndf = tf.size(exp).numpy() - self.normalize
-
-            aux.append(self.chi2(res, rescov).numpy())  # chi2val
-            aux.append(tf.size(exp).numpy() - self.normalize)  # ndf
-        else:
-            aux.append(None)
-            aux.append(None)
-
-        return exp, aux
-
-    def expected_events_projection(
-        self,
-        channel,
-        axes,
-        inclusive=True,
-        compute_variance=True,
-        compute_cov=False,
-        compute_global_impacts=False,
-        compute_variations=False,
-        correlated_variations=False,
-        profile=True,
-        compute_chi2=False,
-        masked=False,
-    ):
-
-        def fun():
-            return self._compute_yields(
-                inclusive=inclusive,
+            pred, pred_var, pred_cov, _1, _2 = self.expected_with_variance(
+                model.make_fun(flat_fun, self.x, inclusive),
                 profile=profile,
-                full=masked,
-            )
-
-        info = self.indata.channel_info[channel]
-        start = info["start"]
-        stop = info["stop"]
-
-        channel_axes = info["axes"]
-
-        exp_axes = channel_axes.copy()
-        hist_axes = [axis for axis in channel_axes if axis.name in axes]
-
-        if len(hist_axes) != len(axes):
-            raise ValueError(
-                f"Hist axes {[h.name for h in hist_axes]} != {axes} not found"
+                compute_cov=True,
             )
 
-        extra_axes = []
-        if not inclusive:
-            exp_axes.append(self.indata.axis_procs)
-            hist_axes.append(self.indata.axis_procs)
-            extra_axes.append(self.indata.axis_procs)
+            chi2val = self.chi2(pred - data, pred_cov + data_cov).numpy()
+            ndf = tf.size(pred).numpy() - getattr(model, "normalize", False)
```

Comment:
Right, rather than using the normalize attribute here, the model should just have an attribute specifying how many degrees of freedom to subtract.